RiotSwiftUI/Modules/Authentication/VerifyMsisdn/Coordinator/AuthenticationVerifyMsisdnCoordinator.swift

```diff
@@ -0,0 +1,220 @@
+            
+            switch result {
+            case .send(let phoneNumber):
+                self.sendSMS(phoneNumber)
+            case .submitOTP(let otp):
+                self.submitOTP(otp)
+            case .resend:
+                self.resendSMS()
+            case .cancel:
+                self.callback?(.cancel)
+            case .goBack:
+                self.authenticationVerifyMsisdnViewModel.goBackToMsisdnForm()
+            }
+        }
+    }
+    
+    /// Show an activity indicator whilst loading.
+    @MainActor private func startLoading() {
+        loadingIndicator = indicatorPresenter.present(.loading(label: VectorL10n.loading, isInteractionBlocking: true))
+    }
+    
+    /// Hide the currently displayed activity indicator.
+    @MainActor private func stopLoading() {
+        loadingIndicator = nil
+    }
+    
+    /// Sends a validation email to the supplied address and then begins polling the server.
+    @MainActor private func sendSMS(_ phoneNumber: String) {
+        startLoading()
+        
+        currentTask = Task { [weak self] in
+            do {
+                let countryCode = try countryCodeFromPhoneNumber(phoneNumber)
+                let threePID = RegisterThreePID.msisdn(msisdn: phoneNumber, countryCode: countryCode)
+                let result = try await registrationWizard.addThreePID(threePID: threePID)
+                
+                // Shouldn't be reachable but just in case, continue the flow.
+                
+                guard !Task.isCancelled else { return }
+                
+                self?.callback?(.completed(result))
+                self?.stopLoading()
+            } catch RegistrationError.waitingForThreePIDValidation {
+                // If three PID validation is required, show OTP screen.
+                authenticationVerifyMsisdnViewModel.updateForSentSMS()
+                self?.stopLoading()
+            } catch is CancellationError {
+                return
+            } catch {
+                self?.stopLoading()
+                self?.handleError(error)
+            }
+        }
+    }
+
+    @MainActor private func submitOTP(_ otp: String) {
+        startLoading()
+
+        currentTask = Task { [weak self] in
+            do {
+                let result = try await registrationWizard.handleValidateThreePID(code: otp)
+                
+                // Shouldn't be reachable but just in case, continue the flow.
+
+                guard !Task.isCancelled else { return }
+
+                self?.callback?(.completed(result))
+                self?.stopLoading()
+            } catch RegistrationError.threePIDClientFailure {
+                self?.stopLoading()
+                self?.handleError(RegistrationError.threePIDClientFailure)
+            } catch RegistrationError.threePIDValidationFailure {
+                self?.stopLoading()
+                self?.handleError(RegistrationError.threePIDValidationFailure)
```

Comment:
As these are all handled in the same place (and end up showing the same error to the user), we can probably remove them.